x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformTask.java

```diff
@@ -505,7 +505,7 @@ public void fail(String reason, ActionListener<Void> listener) {
                 return;
             }
 
-            logger.error("[{}] transform has failed; experienced: [{}].", transform.getId(), reason);
+            logger.atError().withThrowable(exception).log("[{}] transform has failed; experienced: [{}].", transform.getId(), reason);
```

Comment:
Adding just the exception to the original code:
```
logger.error("[{}] transform has failed; experienced: [{}].", transform.getId(), reason, exception);
```
triggers the `ESLoggerUsageChecker`, hence this more verbose approach.

I think this is a shortcoming of the checker, see also:
https://www.slf4j.org/faq.html#paramException

Reply:
If I remember correctly that's not a shortcoming of the checker. The exception doesn't get logged with the simpler syntax because it ends up calling [this method](https://github.com/apache/logging-log4j2/blob/24f4fa990278a2346e8345d8fd4c09246451bd34/log4j-api/src/main/java/org/apache/logging/log4j/spi/AbstractLogger.java#L664) so the exception ends up as an unused message parameter.

The other way people work around this is:

```
logger.error(Strings.format("[{}] transform has failed; experienced: [{}].", transform.getId(), reason), exception);
```

Reply:
The following code:
```
        Exception e = new ElasticsearchException("EXCEPTION");
        Logger logger = LogManager.getLogger(Elasticsearch.class);
        logger.error("zero params + exception", e);
        logger.error("one param [{}] + exception", 42, e);
        logger.error("two params [{}] [{}]", 42, e);
        logger.error("two params + exception [{}] [{}]", 42, e, e);
```

results in:

```
» [2023-11-16T10:29:26,024][ERROR][o.e.b.Elasticsearch      ] [runTask-0] zero params + exception org.elasticsearch.ElasticsearchException: EXCEPTION
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase1(Elasticsearch.java:143)
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:69)
»  
» [2023-11-16T10:29:26,043][ERROR][o.e.b.Elasticsearch      ] [runTask-0] one param [42] + exception org.elasticsearch.ElasticsearchException: EXCEPTION
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase1(Elasticsearch.java:143)
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:69)
»  
» [2023-11-16T10:29:26,044][ERROR][o.e.b.Elasticsearch      ] [runTask-0] two params [42] [org.elasticsearch.ElasticsearchException: EXCEPTION]
» [2023-11-16T10:29:26,044][ERROR][o.e.b.Elasticsearch      ] [runTask-0] two params + exception [42] [org.elasticsearch.ElasticsearchException: EXCEPTION] org.elasticsearch.ElasticsearchException: EXCEPTION
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.initPhase1(Elasticsearch.java:143)
»  	at org.elasticsearch.server@8.12.0-SNAPSHOT/org.elasticsearch.bootstrap.Elasticsearch.main(Elasticsearch.java:69)
```

As you can see: an additional trailing param of the type exception is always logged.

Reply:
This is the behavior since SLF4J 1.6.0 (dating back till 2010). Before that, the exception would be ignored in these cases, see the link above.

Reply:
Regarding:
```
logger.error(Strings.format("[{}] transform has failed; experienced: [{}].", transform.getId(), reason), exception);
```

That works indeed, but has some disadvantages:
- Some logging implementations collect the unexpanded pattern-string as a queryable field.
- It also prevents spending time rendering a message that no logger is configured to output.

(just quoting https://google.github.io/styleguide/pyguide.html#3101-logging)

Reply:
That's interesting. It does look like `ESLoggerUsageChecker` is out-of-date. I know back in about 2017 we were having problems with exceptions getting lost from log statements. Elasticsearch was using log4j version 1 then I think though, so that was probably why. I guess the checker didn't get upgraded when log4j was made more intuitive.

Reply:
Any advice on what to do next?

- file bug report on ESLoggerUsageChecker (and where?)
- do nothing
- fix it myself